File: src/console/views/telemetry.ts
import type { LogCategory, LogLevel } from "../../application/logging";
import type { RequestRoutingMetadata, UsageDimension, UsagePeriod } from "../../application/contracts";
import type { ChartBucket, ModelTokenTotalsRow, UsageByRow, UsageCacheSummary } from "../../storage";


// ---------------------------------------------------------------------------
// Runtime metadata (read-only)
// ---------------------------------------------------------------------------

export interface RequestHistoryFilters {
  readonly period?: UsagePeriod;
  readonly providerId?: string;
  readonly model?: string;
  readonly apiKeyId?: string;
  readonly status?: "ok" | "error";
  readonly limit?: number;
  readonly cursor?: string;
  readonly clientIp?: string;
}

export interface RequestHistoryRow {
  readonly requestId: string;
  readonly endpoint: string;
  readonly surface: string;
  readonly apiKeyId: string | null;
  readonly apiKeyPrefix: string | null;
  readonly clientIp?: string | null;
  readonly providerId: string | null;
  readonly model: string | null;
  readonly statusCode: number;
  readonly errorKind: string | null;
  readonly mode: "non_stream" | "stream";
  readonly startedAt: string;
  readonly finishedAt: string;
  readonly durationMs: number;
  readonly inputTokens: number | null;
  readonly outputTokens: number | null;
  readonly cachedTokens: number | null;
  readonly cacheWriteTokens: number | null;
  readonly reasoningTokens: number | null;
  readonly totalTokens: number | null;
  readonly usageSource: string;
  readonly clientName: string;
  readonly clientSource: string;
  readonly routing: RequestRoutingMetadata;
  readonly messageCount: number;
  readonly toolCount: number;
  readonly imageCount: number;
  readonly tfftMs: number | null;
  readonly payloads?: {
    readonly clientRequest: { readonly text: string; readonly truncated: boolean; readonly originalBytes: number; readonly capturedBytes: number } | null;
    readonly providerRequest: { readonly text: string; readonly truncated: boolean; readonly originalBytes: number; readonly capturedBytes: number } | null;
    readonly providerResponse: { readonly text: string; readonly truncated: boolean; readonly originalBytes: number; readonly capturedBytes: number } | null;
    readonly clientResponse: { readonly text: string; readonly truncated: boolean; readonly originalBytes: number; readonly capturedBytes: number } | null;
  } | null;
}

export interface UsageSummaryView {
  readonly requests: number;
  readonly inputTokens: number;
  readonly outputTokens: number;
  readonly cachedTokens: number;
  readonly totalTokens: number;
  readonly errors: number;
  readonly avgDurationMs: number;
  readonly estimatedCostUsd: number;
  readonly partial: boolean;
}

export interface ProviderTodayView {
  readonly providerId: string;
  readonly requests: number;
  readonly inputTokens: number;
  readonly cachedTokens: number;
  readonly outputTokens: number;
  readonly errors: number;
}

export interface IpSummaryView {
  readonly ip: string;
  readonly requests: number;
  readonly errors: number;
  readonly lastRequestAt: string;
  readonly inputTokens: number;
  readonly outputTokens: number;
}

export interface ConsoleLogLine {
  readonly id: number;
  readonly ts: string;
  readonly level: LogLevel;
  readonly scope: string;
  readonly category: LogCategory;
  readonly msg: string;
}

/** Compact model-probe metadata — never prompt, thinking, body, tool, or image content. */
export interface ModelProbeMetadata {
  readonly providerId: string;
  readonly model: string;
  readonly credentialMode: "auto" | "account" | "manual";
  readonly ok: boolean;
  readonly mode: "stream" | "non_stream" | null;
  readonly latencyMs: number;
  readonly errorKind: string | null;
  readonly occurredAt: string;
}

export interface RuntimeMetadataRepository {
  queryRequests(filters: RequestHistoryFilters): Promise<{ readonly items: readonly RequestHistoryRow[]; readonly nextCursor: string | null }>;
  getRequest(requestId: string): Promise<RequestHistoryRow | null>;
  queryUsageSummary(period: UsagePeriod): Promise<UsageSummaryView>;
  queryUsageCache(period: UsagePeriod): Promise<UsageCacheSummary>;
  queryUsageChart(period: UsagePeriod): Promise<readonly ChartBucket[]>;
  queryUsageBy(dimension: UsageDimension, period: UsagePeriod): Promise<readonly UsageByRow[]>;
  queryModelTokenTotals(period: UsagePeriod): Promise<readonly ModelTokenTotalsRow[]>;
  queryProviderToday(): Promise<readonly ProviderTodayView[]>;
  queryLastProviderError(providerId: string): Promise<string | null>;
  queryIpSummary(limit: number): Promise<readonly IpSummaryView[]>;
  sumKeyTokens(keyId: string): Promise<{ readonly dailyUsed: number; readonly allTimeUsed: number }>;
  queryLogs(limit: number): Promise<readonly ConsoleLogLine[]>;
  clearLogs(): Promise<void>;
  recordModelProbe(meta: ModelProbeMetadata): Promise<void>;
}

// ---------------------------------------------------------------------------

File: daemon/internal/database/models/telemetry.go
package models

import "time"

// ShareUsage is a bounded aggregate used by the public monitor boundary.
// It intentionally contains no client IP, provider, or model dimensions.
type ShareUsage struct {
	TotalRequests   int64
	TotalTokens     int64
	DailyRequests   int64
	DailyTokens     int64
	MonthlyRequests int64
	MonthlyTokens   int64
}

// RequestHistory is one proxy invocation record (request_history).
type RequestHistory struct {
	ID               int64
	TraceID          string
	Endpoint         string
	Surface          string
	APIKeyID         string
	APIKeyPrefix     string
	Provider         string
	Model            string
	Status           int
	ErrorKind        string
	Stream           bool
	StartedAt        time.Time
	FinishedAt       time.Time
	DurationMs       int
	InputTokens      *int
	OutputTokens     *int
	CachedTokens     *int
	CacheWriteTokens *int
	ReasoningTokens  *int
	TotalTokens      *int
	UsageSource      string
	MetaJSON         []byte
	ClientName       string
	ClientSource     string
	MessageCount     int
	ToolCount        int
	ImageCount       int
	TFFTMs           *int
	ClientIP         string
}

// RequestPayload holds captured request/response bodies for a trace
// (request_payloads).
type RequestPayload struct {
	RequestID            string
	ClientRequest        []byte
	ProviderRequest      []byte
	ProviderResponse     []byte
	ClientResponse       []byte
	ClientRequestMeta    []byte
	ProviderRequestMeta  []byte
	ProviderResponseMeta []byte
	ClientResponseMeta   []byte
	CreatedAt            time.Time
	UpdatedAt            time.Time
}

// ConsoleLog is an operator-visible structured log entry (console_logs).
type ConsoleLog struct {
	ID      int64
	TS      time.Time
	Level   string
	Scope   string
	Message string
}
